Utilities/mkpack.py

```diff
@@ -0,0 +1,307 @@
+        
+        def sourcedesc(self):
+            return "resource ID %d from pack %s" % (self.resid, self.resfile)
+    
+    class ResourceFile(Resource):
+        def __init__(self, coll, j):
+            super(self.__class__, self).__init__(coll, j)
+            
+            self.file = "%s/%s" % (self.coll.root, j["input"]["file"], )
+        
+        def deps(self):
+            return [self.file]
+        
+        def data(self):
+            return load_resource_from_disk(self.file)
+        
+        def sourcedesc(self):
+            return self.file
+    
+    def __init__(self, fname, root = "."):
+        """
+        Load in a resource collection from a file, but don't load the
+        resources associated with it.  (That happens later.)
+        """
+        
+        self.jfname = fname
+        self.root = root
+        
+        with open(fname, 'r') as fd:
+            jdb = json.load(fd)
+        
+        if "references" in jdb:
+            self.references = {k: "%s/%s" % (self.root, v) for k, v in jdb["references"].iteritems()}
+        else:
+            self.references = {}
+        
+        self.resources = []
+        for res in jdb["resources"]:
+            if res["input"]["type"] == "file":
+                self.resources.append(ResourceCollection.ResourceFile(self, res))
+            elif res["input"]["type"] == "resource":
+                self.resources.append(ResourceCollection.ResourceRef(self, res))
+            else:
+                raise ValueError("unknown resource type %s" % (res["type"], ))
+    
+    def deps(self):
+        """
+        List of all dependencies of this resource pack.
+        """
+        
+        deps = {}
+        for r in self.resources:
+            for d in r.deps():
+                deps[d] = True
+        return [d for d in deps]
+    
+    def rsrcs(self):
+        """
+        List of raw resource data in this resource pack.
+        """
+        
+        return [r.data() for r in self.resources]
+    
+    def write_pbpack(self, fname):
+        """
+        Write out a .pbpack file with all of the resources loaded.
+        """
+        
+        return save_pbpack(fname, self.rsrcs())
+    
+    def write_header(self, fname):
+        """
+        Write out a C header file that has symbolic names for resource
+        identifiers.
+        """
+        
+        with open(fname, 'w') as fd:
+            fd.write("/* THIS FILE IS AUTOMATICALLY GENERATED BY mkpack.py. */\n")
+            fd.write("/* IF YOU MODIFY IT, YOU WILL BE VERY SAD. */\n")
+            fd.write("/* IF YOU CHECK IT IN, I WILL BE VERY SAD. */\n")
+            fd.write("\n")
+            fd.write("#pragma once\n")
+            fd.write("\n")
+            fd.write("typedef enum resource_id {\n")
+            for (id, r) in enumerate(self.resources):
+                fd.write("    %s = %d, /* (from %s) */\n" % (r.name, id + 1, r.sourcedesc(), ))
+            fd.write("} resource_id;\n")
```

Comment:
Consider using a single triple quoted string (""") for this.

Reply:
I think, specifically, this might make sense for the header (not so much the footer). That said, it does tend to either mess with code flow (you have to slam the string after the first line against the left edge) or add spurious complexity (`textwrap.dedent`), so I have no strong opinions on this.

Reply:
Yeah, I was looking to see if there was a more enjoyable way to solve this, and ran into the same issues Katharine found.  I figured that for the seven lines, I would put up with it, but if it were to be any longer, I'd need to either triple-quote it -- or use a real templating system.